Fixed #26120 -- Made HStoreField cast keys & values to strings

gchp opened the pull request:
The HStoreField will now convert all keys & values to string before
being saved to the database.

timgraham commented on django/contrib/postgres/fields/hstore.py:
```diff
@@ -51,6 +51,20 @@ def formfield(self, **kwargs):
+            prep_value = {}
+            for key, val in value.items():
+                if not isinstance(key, six.string_types):
+                    key = str(key)
```
Could you add some tests with non-ASCII values? I guess we probably want to use `force_text()` here instead of casting to bytes on Python 2.

timgraham commented on django/contrib/postgres/fields/hstore.py:
```diff
@@ -51,6 +51,20 @@ def formfield(self, **kwargs):
+        if isinstance(value, dict):
+            prep_value = {}
+            for key, val in value.items():
+                if not isinstance(key, six.string_types):
```
I wonder if the `isinstance()`condition  results in any performance savings? I tend to think always casting might be simpler.

timgraham commented:
Does `get_prep_value()` also affect queries like `HStoreModel.objects.filter(field__a=1)`? Might be worth testing that.

gchp commented:
@timgraham - thanks for the feedback. I've updated to use `force_text`, and removed the `isinstance` check.

The filter you mention seems to work just fine. I added a test anyway.

mjtamlyn commented on django/contrib/postgres/fields/hstore.py:
```diff
@@ -51,6 +52,19 @@ def formfield(self, **kwargs):
+    def get_prep_value(self, value):
+        value = super(HStoreField, self).get_prep_value(value)
+
+        if isinstance(value, dict):
```
I'd consider changing this to `if value is not None` - I don't know whether other mappables can be handled lower down the stack? It's an edge case though and may be irrelevant as higher levels would have already rejected it.

gchp replied:
Not sure that will work. `get_prep_value` is used for querying too, so if I remove the test for `dict`, this would raise an error:

    HStoreModel.objects.filter(field__has_keys=['a', 'c'])

In this case, `value` is a list.

mjtamlyn replied:
Makes sense

gchp replied:
@mjtamlyn actually, what about this case:

     HStoreModel.objects.get(field__has_keys=[2, 'a'])

I should probably make it convert list elements to strings too?

timgraham commented on tests/postgres_tests/test_hstore.py:
```diff
@@ -37,6 +39,17 @@ def test_value_null(self):
 
+    def test_key_val_cast_to_string(self):
+        value = {'a': 1, 'b': 'B', 2: 'c', 'd': 'ê'}
+        expected_value = {'a': '1', 'b': 'B', '2': 'c', 'd': force_text('ê')}
```
I think `force_text()` isn't needed but rather adding `from __future__ import unicode_literals` to this file.

timgraham replied:
How about a non-ASCII key too, just to be safe?

timgraham replied:
We should check the behavior with bytestrings too. I'm not sure how PostgreSQL stores them and if the change to force them to unicode will have any practical consequences.

gchp replied:
> We should check the behavior with bytestrings too.

@timgraham you mean something like `{'test': b'test'}`?

timgraham replied:
Yes -- does this change affect that case.

gchp replied:
It doesn't seem to, but I'll add it to the test case anyway.

mjtamlyn commented:
LGTM

gchp commented:
@mjtamlyn @timgraham - I've updated the patch to also convert list elements to strings if a list is given to `get_prep_value`. This covers the case where you have a query like so:

    HStoreModel.objects.get(field__has_keys=[2, 'a'])

Let me know if you have any further comments. Thanks for the feedback this far!

mjtamlyn commented:
Looks like tests fail on Py3 now.

gchp commented:
@mjtamlyn that's what I get for being lazy and not checking py3 myself first :/

Updated!

timgraham commented on django/contrib/postgres/fields/hstore.py:
```diff
@@ -51,6 +52,22 @@ def formfield(self, **kwargs):
+            value = prep_value
+
+        if isinstance(value, list):
+            value = [str(item) for item in value]
```
Why `str` instead of `force_text()`?

timgraham commented:
Should the behavior be documented in `docs/ref/contrib/postgres/fields.txt`?

gchp commented:
@timgraham changed str to force_text. Will update the docs now.

mjtamlyn commented:
I don't think we need to document the behaviour, it's equivalent to other fields.

timgraham commented:
merged in 8dea9f089dcb8aea2c417c91344c4e680431e972, thanks!
